File: src/config/Blueprint.php
<?php

namespace jmw\fabricad\config;

use jmw\fabricad\blocks\BasicBuildingBlock;
use jmw\fabricad\shapes\Container;

class Blueprint implements \Iterator
{
    /**
     *
     * @var string
     */
    private $name = '';
    

    /**
     * 
     * @var string
     */    
    private $id = '';
    
    /**
     *
     * @var string
     */
    private $description = '';
    
    
    private $settings = array();
    
    /**
     * 
     * @var array
     */
    private $blocks = array();
    
    public function __construct(string $name = '', string $description = '', $config = array(), $blocks = array())
    {
        $this->setName($name);
        $this->setDescription($description);
        $this->setSettings($config);
        $this->setBlocks($blocks);
    }
    
    /**
     * 
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * 
     * @param string $name
     * @return Blueprint
     */
    public function setName(string $name): Blueprint
    {
        $this->name = $name;
        return $this;
    }
    
    /**
     * Gets the id
     * 
     * @return string
     */
    public function getId(): string
    {
        return $this->id;
    }
    
    /**
     * Sets the id
     * 
     * @param string $id
     * @return Blueprint
     */
    public function setId(string $id): Blueprint
    {
        $this->id = $id;
        return $this;
    }

    /**
     * 
     * @return string
     */
    public function getDescription(): string
    {
        return $this->description;
    }

    /**
     * 
     * @param string $description
     * @return Blueprint
     */
    public function setDescription(string $description): Blueprint
    {
        $this->description = $description;
        return $this;
    }

    /**
     * 
     * @return BasicBuildingBlock[]
     */
    public function getBlocks(): array
    {
        return $this->blocks;
    }
    
    public function addBlock(BasicBuildingBlock $block): Blueprint
    {
        $this->blocks[] = $block;
        
        return $this;
    }
    
    /**
     * Sets the settings to this array
     * 
     * @return array
     */
    public function getSettings(): array
    {
        return $this->settings;
    }
    
    /**
     * 
     * @param array $settings
     * @return Blueprint
     */
    public function setSettings(array $settings): Blueprint
    {
        $this->settings = $settings;
        return $this;
    }

    /**
     * 
     * @param AbstractBuildingBlock[] $blocks
     * @return Blueprint
     */
    public function setBlocks(array $blocks): Blueprint
    {
        $this->blocks = $blocks;
        return $this;
    }
    
    public function size(): int
    {
        return count($this->blocks);
    }
    
    
    public function render(): Container {
        $c = new Container();
        
        echo "\nstart".date('U')."\n";

        // we walk backward!
        for($i = $this->size() -1; $i >= 0 ; $i--) {
            $shape = $this->blocks[$i]->render();
            if ($shape instanceof Container) {
                $items = $shape->flatten();
            } else {
                $items = [$shape];
            }
            
            foreach($items as $item) {
                echo 'a';
                $c->addNonOverlappingParts($item);
                echo 'b';
                //$c->addShape($item);
            }
        }
        echo "\nend".date('U')."\n";
        
        return $c;
    }
    
    // -------------------------------------------------------------------------
    // Iterator implementation
    private $it_counter = 0;
    
    public function next()
    {
        ++$this->it_counter;
    }
    
    public function valid()
    {
        return isset($this->blocks[$this->it_counter]);
    }
    
    public function current()
    {
        return $this->blocks[$this->it_counter];
    }
    
    public function rewind()
    {
        $this->it_counter = 0;
    }
    
    public function key()
    {
        return $this->it_counter;
    }
    
    
    
}

File: src/config/Blueprint.cc

#include <iostream>

#include "Blueprint.h"

namespace fabricad::config
{

  std::string Blueprint::getId()
  {
    return id_;
  }

  Blueprint* Blueprint::setId(std::string id)
  {
    id_ = id;
    return this;
  }

  std::string Blueprint::toString()
  {
    std::string s =
      "id         : " + getId()          + "\n" +
      "Name       : " + getName()        + "\n" +
      "Description: " + getDescription() + "\n" +
      "Blocks     :\n";

    for(size_t i = 0 ; i < blocks_.size() ; ++i)
    {
      s += blocks_[i]->toString("    ");
      s += "    ----------\n";
    }

    return s;
  }

  std::vector<BasicBuildingBlock*> Blueprint::getBlocks()
  {
    return blocks_;
  }

  size_t Blueprint::getSize()
  {
    return blocks_.size();
  }

  Blueprint* Blueprint::addBlock(BasicBuildingBlock* block)
  {
    blocks_.push_back(block);
    return this;
  }

/*
  shapelayer Blueprint::getLayer(size_t layer)
  {
    if (layers_.empty()) {
      render();
    }

    return layers_.at(layer);
  }
*/
  std::map<BasicBuildingBlock*, std::vector<shapelayer>> Blueprint::getLayers()
  {
    if (shapeElements_.empty()) {
      render();
    }
    return shapeElements_;
  }

  void Blueprint::render()
  {
    shapeElements_.clear();

    std::vector<polygon> tocheck;

    // Walk the blocks, and add all shapes
    // We walk the list backwards, so that we can check with the
    // previous shapes what to exclude
    // for(size_t index = 0 ; index < blocks_.size() ; index++)
    for(auto& block: blocks_)
    {
      // size_t index = blocks_.size() -1 - i;
      // std::cout << "Working on: " << block->getName() << std::endl;
      std::vector<shapelayer> layers;
      initializeLayerSet(layers);

      for(size_t layer = 0 ; layer < 4 ; layer++) {

        size_t l = 3 - layer;
        if (!block->getLayer(l).lines.empty()) {
          // here we want to calculate the difference of each element
          // with the elements of the first layer, as that layer contains
          // the outside shapes.
          std::vector<linestring> lines;
          calculateDifference(block->getLayer(l).lines, tocheck, &lines);
          linestringmerge(&layers[l].lines, lines);
        }

        if (!block->getLayer(l).polygons.empty()) {
          // Polygons we always add, independent of whether they overlap
          // or not.
          polygonmerge(&layers[l].polygons, block->getLayer(l).polygons);

          if (l == 0) {
            polygonmerge(&tocheck, block->getLayer(l).polygons);
          }
        }
      }
      // Add block-layers pair to the shapeElements
      shapeElements_.insert( std::pair<BasicBuildingBlock*, std::vector<shapelayer>>(block, layers));
    }
  }

  Blueprint::~Blueprint()
  {
    // Remove all blueprints
    size_t size = blocks_.size();
    for(size_t i = 0 ; i < size; ++i)
    {
      delete blocks_[i];
    }
  }

  void Blueprint::initializeLayerSet(std::vector<shapelayer> &layers)
  {
    shapelayer shapes;
    shapes.id = "shapes";
    shapes.name = "Shapes";
    layers.push_back(shapes);

    shapelayer bricks;
    bricks.id = "bricks";
    bricks.name = "Brickwork";
    layers.push_back(bricks);

    shapelayer cutout;
    cutout.id = "cutout";
    cutout.name = "Cutouts";
    layers.push_back(cutout);

    shapelayer other;
    other.id = "other";
    other.name = "Other shapes";
    layers.push_back(other);
  }
}
